### ISSUE TITLE
Generate sample from categorical pandas dataframe

### ISSUE THREAD

**Issue by mjuarezm** [pgmpy/pgmpy#1244]
### Subject of the issue
There seems to be an issue when using pandas dataframes to create a Bayesian model and then sample from it. It looks like `pgmpy` does not create the indices of the CPD tables correctly when the contents of the dataframe are categorical.

### Your environment
* pgmpy version: 0.1.10
* Python version: 3.6
* Operating System: arch

### Steps to reproduce
I have a dataframe with two variables `type` and `sex` that take categorical values. 

type | sex
-- | --
manager | m
manager | f
manager | m
senior_executive | m

I create a simple Bayesian model like this:
```
G = BayesianModel([('sex', 'type')])
G.fit(df)
G.check_model()   # check passes
for cpd in G.get_cpds():
    print(cpd)
```

sex(f) | 0.296017
-- | --
**sex(m)** | **0.703983** 

sex                    | sex(f)               | sex(m)
-- | -- |  --
**type(executive)**        | 0.011741682974559686 | 0.018103270931907015
**type(experienced)**      | 0.017123287671232876 | 0.019131865871219914
**type(manager)**          | 0.7059686888454012   | 0.5856819584447645
**type(senior_executive)** | 0.12181996086105674  | 0.3112528286360831
**type(student)**          | 0.14334637964774952  | 0.06583007611602551 

when I try to sample from it, I get a `KeyError`:

```
inf = BayesianModelSampling(G)
inf.forward_sample(size=5)
```

This is the error message:
> ---------------------------------------------------------------------------
> KeyError                                  Traceback (most recent call last)
> <ipython-input-63-8b65c8622f09> in <module>
> ----> 1 inf.forward_sample(size=5)
> 
> /usr/lib/python3.8/site-packages/pgmpy/sampling/Sampling.py in forward_sample(self, size, return_type)
>      81             evidence = cpd.variables[:0:-1]
>      82             if evidence:
> ---> 83                 cached_values = self.pre_compute_reduce(variable=node)
>      84                 evidence = np.vstack([sampled[i] for i in evidence])
>      85                 weights = list(map(lambda t: cached_values[tuple(t)], evidence.T))
> 
> /usr/lib/python3.8/site-packages/pgmpy/sampling/Sampling.py in pre_compute_reduce(self, variable)
>      99         ):
>     100             states = list(zip(variable_evid, state_combination))
> --> 101             cached_values[state_combination] = variable_cpd.reduce(
>     102                 states, inplace=False
>     103             ).values
> 
> /usr/lib/python3.8/site-packages/pgmpy/factors/discrete/CPD.py in reduce(self, values, inplace)
>     357         tabular_cpd = self if inplace else self.copy()
>     358 
> --> 359         super(TabularCPD, tabular_cpd).reduce(values)
>     360         tabular_cpd.normalize()
>     361 
> 
> /usr/lib/python3.8/site-packages/pgmpy/factors/discrete/DiscreteFactor.py in reduce(self, values, inplace)
>     425 
>     426         phi = self if inplace else self.copy()
> --> 427         values = [
>     428             (var, self.get_state_no(var, state_name)) for var, state_name in values
>     429         ]
> 
> /usr/lib/python3.8/site-packages/pgmpy/factors/discrete/DiscreteFactor.py in <listcomp>(.0)
>     426         phi = self if inplace else self.copy()
>     427         values = [
> --> 428             (var, self.get_state_no(var, state_name)) for var, state_name in values
>     429         ]
>     430 
> 
> /usr/lib/python3.8/site-packages/pgmpy/utils/state_name.py in get_state_no(self, var, state_name)
>      72         """
>      73         if self.state_names:
> ---> 74             return self.name_to_no[var][state_name]
>      75         else:
>      76             return state_name
> 
> KeyError: 0

I think the issues is that, internally, `pgmpy` is storing the variables as dictionaries. 

```
G.get_cpds()[1].name_to_no
```

> {'type': {'executive': 0,
>   'experienced': 1,
>   'manager': 2,
>   'senior_executive': 3,
>   'student': 4},
>  'sex': {'f': 0, 'm': 1}}

### Expected behaviour
I think `pgmpy` should factorize the variable names itself but it does not.

### Actual behaviour
The variable values are treated as dictionaries, throwing an error on sampling.

**Comment by ankurankan** [comment]
@mjuarezm Thanks for reporting this. This should be working on the latest dev. There was a bug in the reduce method and has been fixed now. But the sampling method right now  returns the state numbers instead of state names.

**Comment by mjuarezm** [comment]
Thanks for the quick reply. I confirm the issue is fixed in version: `0.1.10dev`
